src/Sylius/Bundle/ShopBundle/Tests/Doctrine/ORM/ProductRepositoryTest.php

```diff
@@ -0,0 +1,80 @@
+<?php
+/*
+ * This file is part of the Sylius package.
+ *
+ * (c) Paweł Jędrzejewski
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\ShopBundle\Tests\Doctrine\ORM;
+
+use Doctrine\ORM\Mapping\ClassMetadata;
+use Sylius\Bundle\CoreBundle\Doctrine\ORM\ProductRepository;
+use Sylius\Component\Core\Model\Product;
+
+final class ProductRepositoryTest extends DoctrineRepositoryTestCase
+{
+    /** @var ProductRepository */
+    private $productRepository;
+
+    private function createProductRepository(): ProductRepository
+    {
+        return new ProductRepository($this->entityManager, new ClassMetadata(Product::class));
+    }
+
+    /**
+     * @test
+     */
+    public function it_finds_all_products_with_given_taxon_when_flag_is_not_set(): void
+    {
+        $childTaxon = $this->createTaxon('sampleChildTaxon');
+        $parentTaxon = $this->createTaxon('sampleParentTaxon', [$childTaxon]);
+
+        $locale = $this->createLocale('en_US');
+        $currency = $this->createCurrency('USD');
+        $channel = $this->createChannel('sampleChannel', $locale, $currency);
+
+        $firstProduct = $this->createProduct('firstProduct', [$channel], [$parentTaxon]);
+        $secondProduct = $this->createProduct('secondProduct', [$channel], [$childTaxon]);
+
+        $productRepository = $this->createProductRepository();
+
+        $result = $productRepository
+            ->createShopListQueryBuilder($channel, $parentTaxon, 'en_US', [], false)
+            ->getQuery()
+            ->getResult();
+
+        self::assertCount(1, $result);
+        self::assertEquals($firstProduct->getCode(), $result[0]->getCode());
+    }
+
+    /**
+     * @test
+     */
+    public function it_finds_all_products_with_given_taxon_and_its_children_when_flag_is_set(): void
+    {
+        $childTaxon = $this->createTaxon('sampleChildTaxon');
+        $parentTaxon = $this->createTaxon('sampleParentTaxon', [$childTaxon]);
+
+        $locale = $this->createLocale('en_US');
+        $currency = $this->createCurrency('USD');
+        $channel = $this->createChannel('sampleChannel', $locale, $currency);
+
+        $this->createProduct('firstProduct', [$channel], [$parentTaxon]);
+        $this->createProduct('secondProduct', [$channel], [$childTaxon]);
+
+        $productRepository = $this->createProductRepository();
+
+        $result = $productRepository
+            ->createShopListQueryBuilder($channel, $parentTaxon, 'en_US', [], true)
+            ->getQuery()
+            ->getResult()
+        ;
+
+        self::assertCount(2, $result);
```

Comment:
Could we add another root taxon with a product in it to test whether there isn't just `findAll()` in the implementation? :)